app/src/main/kotlin/com/wire/android/ui/home/conversations/MessageExpiration.kt

```diff
@@ -0,0 +1,266 @@
+                }
+
+                return timeLeftUpdateInterval
+            }
+
+            fun decreaseTimeLeft(interval: Duration) {
+                if (timeLeft.inWholeSeconds != 0L) timeLeft -= interval
+            }
+
+            fun alphaBackgroundColor(): Float {
+                val totalTimeLeftRatio = timeLeft / expireAfter
+
+                return if (totalTimeLeftRatio >= TIME_LEFT_RATIO_BOUNDARY_FOR_ALMOST_TIME_LEFT_ELAPSED_ALPHA) {
+                    PAST_RATIO_BOUNDARY_FOR_ALMOST_TIME_LEFT_ALPHA_VALUE
+                } else {
+                    BEFORE_RATIO_BOUNDARY_FOR_ALMOST_TIME_LEFT_ALPHA_VALUE
+                }
+            }
+        }
+
+        object NotExpirable : SelfDeletionTimerState()
+    }
+}
+
+@Composable
+fun startDeletionTimer(
+    message: UIMessage,
+    expirableTimer: SelfDeletionTimer.SelfDeletionTimerState.Expirable,
+    onStartMessageSelfDeletion: (UIMessage) -> Unit
+) {
+    message.messageContent?.let {
+        when (val messageContent = message.messageContent) {
+            is UIMessageContent.AssetMessage -> startAssetDeletion(
+                expirableTimer = expirableTimer,
+                onSelfDeletingMessageRead = { onStartMessageSelfDeletion(message) },
+                downloadStatus = messageContent.downloadStatus
+            )
+
+            is UIMessageContent.AudioAssetMessage -> startAssetDeletion(
+                expirableTimer = expirableTimer,
+                onSelfDeletingMessageRead = { onStartMessageSelfDeletion(message) },
+                downloadStatus = messageContent.downloadStatus
+            )
+
+            is UIMessageContent.ImageMessage -> startAssetDeletion(
+                expirableTimer = expirableTimer,
+                onSelfDeletingMessageRead = { onStartMessageSelfDeletion(message) },
+                downloadStatus = messageContent.downloadStatus
+            )
+
+            is UIMessageContent.TextMessage -> {
+                LaunchedEffect(Unit) {
+                    onStartMessageSelfDeletion(message)
+                }
+                LaunchedEffect(expirableTimer.timeLeft) {
+                    with(expirableTimer) {
+                        if (timeLeft != ZERO) {
+                            delay(updateInterval())
+                            decreaseTimeLeft(updateInterval())
+                        }
+                    }
+                }
+            }
+
+            else -> {}
+        }
+    }
+}
+
+@Composable
+private fun startAssetDeletion(
+    expirableTimer: SelfDeletionTimer.SelfDeletionTimerState.Expirable,
+    onSelfDeletingMessageRead: () -> Unit,
+    downloadStatus: Message.DownloadStatus
+) {
+    LaunchedEffect(downloadStatus) {
+        if (downloadStatus == Message.DownloadStatus.SAVED_EXTERNALLY || downloadStatus == Message.DownloadStatus.SAVED_INTERNALLY) {
+            onSelfDeletingMessageRead()
+        }
+    }
```

Comment:
it is the same condition indeed, however the other LaunchedEffect takes timeLeft as a key as well, because we want to start the timer when the asset is downloaded, but also keep launching the LaunchedEffect when the timeLeft changes to keep the timer "ticking" so I do not see a way to move it there 🤔

Reply:
I could move onSelfDeletingMessageRead() to the other one but it would be called each time the timer ticks :S